dawn/test/unit-test/dawn/Optimizer/Passes/TestPassRemoveScalars.cpp

```diff
@@ -0,0 +1,587 @@
+//===--------------------------------------------------------------------------------*- C++ -*-===//
+//                          _
+//                         | |
+//                       __| | __ ___      ___ ___
+//                      / _` |/ _` \ \ /\ / / '_  |
+//                     | (_| | (_| |\ V  V /| | | |
+//                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+//
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+#include "dawn/Compiler/DawnCompiler.h"
+#include "dawn/Optimizer/OptimizerContext.h"
+#include "dawn/Optimizer/PassRemoveScalars.h"
+#include "dawn/Unittest/IIRBuilder.h"
+
+#include <gtest/gtest.h>
+
+using namespace dawn;
+
+namespace {
+
+std::shared_ptr<iir::Expr> getRhsOfAssignment(const std::shared_ptr<iir::Stmt> stmt) {
+  if(stmt->getKind() == iir::Stmt::Kind::VarDeclStmt) {
+    const auto& varDeclStmt = std::dynamic_pointer_cast<iir::VarDeclStmt>(stmt);
+    return varDeclStmt->getInitList()[0];
+  } else if(stmt->getKind() == iir::Stmt::Kind::ExprStmt) {
+    const auto& exprStmt = std::dynamic_pointer_cast<iir::ExprStmt>(stmt);
+    if(exprStmt->getExpr()->getKind() == iir::Expr::Kind::AssignmentExpr) {
+      const auto& assignmentExpr =
+          std::dynamic_pointer_cast<iir::AssignmentExpr>(exprStmt->getExpr());
+      return assignmentExpr->getRight();
+    }
+  }
+
+  return nullptr;
+}
+
+iir::DoMethod& getDoMethod(std::shared_ptr<iir::StencilInstantiation>& si) {
+  auto& iir = si->getIIR();
+  auto& stencil = iir->getChild(0);
+  auto& ms = stencil->getChild(0);
+  auto& stage = ms->getChild(0);
+  return *stage->getChild(0);
+}
+
+std::shared_ptr<iir::Stmt> getNthStmt(std::shared_ptr<iir::StencilInstantiation>& si, int n) {
+  return getDoMethod(si).getAST().getStatements()[n];
+}
+
+bool isVarInDoMethodsAccesses(int varAccessID, const iir::DoMethod& doMethod) {
+  for(const auto& stmt : doMethod.getAST().getStatements()) {
+    const auto& access = stmt->getData<iir::IIRStmtData>().CallerAccesses;
+
+    for(auto& accessPair : access->getWriteAccesses()) {
+      if(varAccessID == accessPair.first) { // first = AccessID
+        return true;
+      }
+    }
+    for(const auto& accessPair : access->getReadAccesses()) {
+      if(varAccessID == accessPair.first) { // first = AccessID
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+TEST(TestRemoveScalars, test_unstructured_scalar_01) {
+  using namespace dawn::iir;
+
+  UnstructuredIIRBuilder b;
```

Comment:
Current state is that local variable types are derived info in the sense that they are not serialized, but they are a necessary input for this pass.

Reply:
ok

Reply:
What we understood is that existing tests from sir/DSL would be changed into serialized iir, not that we would disallow using the IIRBuilder for unit testing passes.
Constructing one of the input IIRs, in this case, is trivially done with the IIRBuilder in less than 10 lines of code.

Concerning local variable types: I think they better fit as derived info because they are always consumed by the passes that run immediately after `PassLocalVarType` and only by them. Same pattern as for the stage graph.
If we have a reason to change that, however, doing it is outside the scope of this PR.